dataflow/flex-templates/pipeline_with_dependencies/README.md

```diff
@@ -0,0 +1,169 @@
+
+# Dataflow flex template: a pipeline with dependencies and a custom container image.
+
+This project is created to illustrate the following setup of a Dataflow Python pipeline:
+- The pipeline is a package that consists of [multiple files](https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/#multiple-file-dependencies).
+- The pipeline has at least one dependency that is not provided in the default Dataflow runtime environment.
+- The workflow uses a [custom container image](https://cloud.google.com/dataflow/docs/guides/using-custom-containers) to preinstall dependencies and to define the pipeline runtime environment.
+- The workflow uses a [Dataflow Flex Template](https://cloud.google.com/dataflow/docs/concepts/dataflow-templates) to control the pipeline submission environment.
+- The runtime and submission environment use same set of Python dependencies and can be created in a reproducible manner.
+
+To illustrate this setup, we use a pipeline that does the following:
+
+1. Finds the longest word in an input file
+2. Creates a [FIGLet text banner](https://en.wikipedia.org/wiki/FIGlet) from of it using [pyfiglet](https://pypi.org/project/pyfiglet/)
+3. Outputs the text banner in another file
+
+
+## The structure of the example
+
+The pipeline package is comprised of the `my_package` directory and the `setup.py` file. The package defines the pipeline, the pipeline dependencies, and the input parameters. You can define multiple pipelines in the same package. The `my_package.launcher` module is used to submit the pipeline to a runner.
+
+The `main.py` file provides a top-level entrypoint to trigger the pipeline launcher from a
+launch environment.
+
+The `Dockerfile` defines the runtime environment for the pipeline. It also configures the Flex Template, which lets you reuse the runtime image to build the Flex Template.
+
+The`Dockerfile.flex_template` and `flex_template_cloudbuild.yaml` files provide the remaining instructions to build the Flex Template.
+
+The `requirements.txt` file defines all Python packages in the dependency chain of the pipeline package. Use it to create reproducible Python environments in the Docker images.
+
+## Before you begin
+
+Follow the
+[Dataflow setup instructions](../../README.md).
+
+## Create a Cloud Storage bucket
+
+```sh
+export PROJECT="project-id"
+export BUCKET="your-bucket"
+export REGION="us-central1"
+gsutil mb -p $PROJECT gs://$BUCKET
+```
+
+## Create an Artifact Registry repository
+
+```sh
+export REPOSITORY="your-repository"
+gcloud artifacts repositories create $REPOSITORY \
+    --repository-format=docker \
+    --location=$REGION
+
+gcloud artifacts repositories create $REPOSITORY \
+    --repository-format=docker \
+    --location=$REGION \
+    --project $PROJECT
+```
```

Comment:
I remember I have to do these as well to get the access:
```
gcloud auth login
gcloud auth application-default login
# replace it with the appropriate region
gcloud auth configure-docker us-docker.pkg.dev
# or if you use docker-credential-gcr
docker-credential-gcr configure-docker --registries=us-docker.pkg.dev
```
Maybe it is different if using `gcloud builds`?

Reply:
> gcloud auth application-default login

This is mentioned in Dataflow prerequisite instructions.

> gcloud auth configure-docker us-docker.pkg.dev

Will add this. It's easy, and it helps if you use local docker.